### PR TITLE
Add function to check for group

### PR DESCRIPTION
**Pull request by florianriche**
**Current content of the review:**

* Modified all the scopes string into AccesstType Enum values
* Added an `authorizations.py` files in `/crud` folder to test groups match and that an access is an admin or not.
  * My current vision would be to simply use the functions defined in this file everywhere it is needed (in routes mostly). I think it would handle most of our use cases we have defined in #137  (except the fetch use case)
  * Added a  ̀groups.py` file in  `/crud` folder to find the group of any entity.
  * ADded corresponding uTest

Admittedly this design is not very clean but to be honest I don't know how else to do it.
(EDITED)

### CODE REVIEW

File: src/app/api/crud/authorizations.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+
+async def are_same_group(group_id_a: int, group_id_b: int) -> bool:
+    if group_id_a != group_id_b:
+        raise HTTPException(status_code=400, detail="You don't belong to the same group")
+    return True
+
+
+async def is_admin_access(access_id: int) -> bool:
+    access = await crud.base.get_entry(accesses, access_id)
```
Comment (frgfm) [inline src/app/api/crud/authorizations.py:32]: Note for later: we only need 1 field, perhaps we should either add the flexibility to specify the fields or write a custom query.

File: src/app/api/crud/authorizations.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+
+async def is_access_in_same_group(access_id: int, group_id: int) -> bool:
+    access = await crud.base.get_entry(accesses, access_id)
+    return await are_same_group(access.group_id, group_id)
+
+
+async def are_same_group(group_id_a: int, group_id_b: int) -> bool:
+    if group_id_a != group_id_b:
+        raise HTTPException(status_code=400, detail="You don't belong to the same group")
```
Comment (frgfm) [inline src/app/api/crud/authorizations.py:27]: 401 I believe ;)
```
from fastapi import status
raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Permission denied")
```

File: src/app/api/crud/authorizations.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    user_or_device = await crud.base.get_entry(table, entry_id)
+    return await is_access_in_same_group(user_or_device.access_id, group_id)
+
+
+async def is_access_in_same_group(access_id: int, group_id: int) -> bool:
+    access = await crud.base.get_entry(accesses, access_id)
+    return await are_same_group(access.group_id, group_id)
+
+
+async def are_same_group(group_id_a: int, group_id_b: int) -> bool:
```
Comment (frgfm) [inline src/app/api/crud/authorizations.py:25]: this doesn't need to be async

File: src/app/api/crud/authorizations.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+
+async def is_in_same_group(table: Table, entry_id: int, group_id: int) -> bool:
+    user_or_device = await crud.base.get_entry(table, entry_id)
+    return await is_access_in_same_group(user_or_device.access_id, group_id)
+
+
+async def is_access_in_same_group(access_id: int, group_id: int) -> bool:
+    access = await crud.base.get_entry(accesses, access_id)
+    return await are_same_group(access.group_id, group_id)
```
Comment (frgfm) [inline src/app/api/crud/authorizations.py:22]: if this is only a value comparison, perhaps the extra function isn't needed?

File: src/app/api/crud/authorizations.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+from app.api.schemas import AccessType
+from app.db import accesses
+
+
+async def is_in_same_group(table: Table, entry_id: int, group_id: int) -> bool:
+    user_or_device = await crud.base.get_entry(table, entry_id)
+    return await is_access_in_same_group(user_or_device.access_id, group_id)
+
+
+async def is_access_in_same_group(access_id: int, group_id: int) -> bool:
```
Comment (frgfm) [inline src/app/api/crud/authorizations.py:20]: I suggest `is_access_in_group` considering the signature

**Review by frgfm — COMMENTED** [review]
Thanks for the PR! I added a few comments

File: src/app/api/crud/authorizations.py (hunk 4 above)
Comment (florianriche) [reply src/app/api/crud/authorizations.py:22]: Its purpose is mainly to raise the HTTPException and to be called directly by objects that contain a group_id directly (like `sites`)
We can do without but I would personally keep it

**Comment by florianriche** [comment]
Thanks for the feedbacks, do you approve this design?

If I end up with additional obstacles I will mention them here.

**Comment by frgfm** [comment]
@florianriche yes thanks for taking care of this!

File: src/app/api/crud/authorizations.py (hunk 2 above)
Comment (florianriche) [reply src/app/api/crud/authorizations.py:27]: Ok, I copy-pasted from antoher part, there are still 400 errors returned with that message :)

**Comment by florianriche** [comment]
Shall we add many scopes in this pull request?

Personnal check-list that I will update.
- [x]  User:  routes all have Admin or can't alter group so no check needed
- [x] Device routes: routes all have Admin or can't alter group so no check needed
- [X] Alerts: routes all have aDmin or can't alter group so no check needed
- [x] Installations: Only Admin scopes
- [x] Sites: Added when requested (and added scopes for GET ressources)
- [x] media: updated  GET resources, Fetch sce
- [X] events: route access
- [x] accesses: only Admin

**Comment by frgfm** [comment]
I think we should keep this PR only about being able to trace the group for each entity. And in another PR implement the next group constraints :ok_hand: (those are hard to review carefully)

**Comment by frgfm** [comment]
For reference, this PR is linked to #125

**Comment by frgfm** [comment]
You got some import conflicts during test collection :sweat_smile:

**Comment by florianriche** [comment]
> You got some import conflicts during test collection

Yes fixed it :)

File: src/app/api/crud/groups.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+
+async def _get_media_group_id(entry_id: int) -> int:
+    entry = await crud.base.get_entry(media, entry_id)
+    return await _get_device_group_id(entry["device_id"])
+
+
+async def _get_installation_group_id(entry_id: int) -> int:
+    entry = await crud.base.get_entry(installations, entry_id)
+    return await _get_device_group_id(entry["device_id"])
```
Comment (frgfm) [inline src/app/api/crud/groups.py:73]: Slight improvement suggestion: it's cheaper using _get_site_group_id in terms of DB operations ;)

File: src/app/api/crud/groups.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+from app.api import crud
+from app.api.schemas import AccessType
+from app.db import (
+    accesses, users, groups,
+    sites, events, devices, media,
+    installations, alerts)
+
+
+async def get_entity_group_id(table: Table, entry_id: int) -> int:
```
Comment (frgfm) [inline src/app/api/crud/groups.py:19]: Improvement suggestion: for cleaner signature, I suggest passing directly the entry. Then the ifs can be checking whether it is `AccessRead` or other reading schemas

File: src/tests/crud/test_authorizations.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+        [2, 1, False],  # Because Admin
+        [2, 2, False],
+        [2, None, False]
+
+    ],
+)
+@pytest.mark.asyncio
+async def test_check_group_update(test_app_asyncio, init_test_db, access_id, group_id, should_raise):
+    if should_raise:
+        with pytest.raises(Exception):
```
Comment (frgfm) [inline src/tests/crud/test_authorizations.py:180]: HTTPException?

File: src/tests/crud/test_authorizations.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+        [1, 1, False],
+        [1, 2, True],
+        [2, 1, False],  # Because Admin
+        [2, 2, False]
+    ],
+)
+@pytest.mark.asyncio
+async def test_check_group_read(test_app_asyncio, init_test_db, access_id, group_id, should_raise):
+    if should_raise:
+        with pytest.raises(Exception):
```
Comment (frgfm) [inline src/tests/crud/test_authorizations.py:159]: HTTPException?

**Review by frgfm — COMMENTED** [review]
Thank you very much! Everything looks good, I only added a few adjustment suggestions :ok_hand:

File: src/app/api/crud/groups.py (hunk 7 above)
Comment (florianriche) [reply src/app/api/crud/groups.py:19]: Not sure it will be more efficient. One will need to retrieve the entity before doing the group_id check

Comment (florianriche) [reply src/app/api/crud/groups.py:19]: But it's true that it is already what we do.
Ok I'll try. But I fear that the Schema checks might be a more brittle design.

**Review by frgfm — APPROVED** [review]
Thanks a lot!

**Pull request merged**